### PR TITLE
feat(DataGrid): Create base implementation

### LINKED ISSUE #2657: Table with selectable rows and also checkbox selection
### Description

- Make the table component more interactive by providing selectable rows and hover styles when rows are selectable. 
- Allow for checkbox selection on rows

I ended up building this out of the partial components of the current table, for a link to the repo contact me via NDS slack

### LINKED ISSUE #3730: DataGrid component
### Description

As recently discussed, there is a common need for an advanced table component that is specifically designed for data.

For our project, the following would be useful (comparing with the Table component):

- **Checkboxes for each row, as per #2657** 

  This includes a select all checkbox in the header, and shading of selected rows. It should also be possible to disable some rows/checkboxes. See also my comment on that issue.

- **Review of font size and use of capitals in the header**

   These both seem inconsistent with other components in the design system. On the font size, it becomes noticeable when comparing with other things on the page, and when you embed other components in a table cell.

- **Review of background colour**

  It currently doesn't have one, so it ends up being transparent. Perhaps it should have a default background colour, and people can override it using Styled Components if they want something different.

- **Perhaps easier control over column widths**

  Currently Table uses `table-layout: fixed`, which results in each column having an equal width by default (not so useful when the contents of some columns are wider than others).

  `table-layout: auto` works okay for us for now. But I can imagine scenarios where you would want `table-layout: fixed` with percentages for some columns (e.g. if you have pagination, you probably don't want the column widths to change when navigating between pages). This could be probably be currently done with the `css` prop on `TableColumn`, but perhaps it could be nicer.

- **Right-aligned columns**

  There may be cases where right-aligned columns (including the heading) are needed e.g. for numbers.

If I can think of anything else will post back here.

### PR DESCRIPTION
**Pull request by m7kvqbe1**
## Related issue

Closes #3730, Closes #2657

## Overview

Create baseline DataGrid that has feature parity with existing Table component. 

Enhance to include some other additional baseline features.

## Reason

Table is lacking the features required to interrogate and manipulate large data sets.

## Work carried out

- [x] Feature parity with Table component (Column sorting, Arbitrary cell content, Caption)
- [x] Add ability to select rows with Checkbox and All Rows (header) + callbacks
- [x] Hover and active state styling for individual rows + callbacks
- [x] Solid background color as opposed to transparent
- [x] Consumer controlled column sizing (https://tanstack.com/table/latest/docs/guide/column-sizing) 
- [x] Enable custom alignment for columns

## Screenshot

Screenshot 2024-04-17 at 08 51 23
Screenshot 2024-04-17 at 08 51 30
Screenshot 2024-04-17 at 08 51 42

## Developer notes

https://github.com/Royal-Navy/design-system/issues/3730

https://github.com/Royal-Navy/design-system/issues/2657

### CODE REVIEW

**Comment by m7kvqbe1** [comment]
@jpveooys @thyhjwb6

File: packages/react-component-library/src/components/DataGrid/DataGrid.stories.tsx
Comment (markhigham) [inline packages/react-component-library/src/components/DataGrid/DataGrid.stories.tsx]: does this need to be exported?

File: packages/react-component-library/src/components/DataGrid/DataGrid.test.tsx
Comment (markhigham) [inline packages/react-component-library/src/components/DataGrid/DataGrid.test.tsx]: :)

File: packages/react-component-library/src/components/DataGrid/DataGrid.test.tsx
Comment (markhigham) [inline packages/react-component-library/src/components/DataGrid/DataGrid.test.tsx]: We use this hack frequently. How about creating a single function that does the same thing. 
I'm thinking that we can then annotate that function with a note to future maintainers explaining why we arrived at this hack

`await hackyWaitForHook()`

```
async function hackyWaitForHook() {
    // Apologies for the hack but we needed to to do this because ...
    return new Promise((resolve) => setTimeout(resolve, 100))
}
```

File: packages/react-component-library/src/components/DataGrid/DataGrid.tsx
Comment (markhigham) [inline packages/react-component-library/src/components/DataGrid/DataGrid.tsx]: maybe add an `index.ts` to `./partials` to simplify these imports

**Review by markhigham — APPROVED** [review]
LGTM: We should add a follow up tickets once #3746 is merged and `mdx` is enabled.
- Add a description mdx file for each of the new storybook folders. 
- Document why Table is deprecated in the mdx
- Add some text to the deprecated stories with links to the docs

**Pull request merged**
